common/src/main/java/com/microsoft/hydralab/common/network/AndroidNetworkMonitor.java

```diff
@@ -0,0 +1,155 @@
+import com.microsoft.hydralab.common.util.ThreadUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.parser.Parser;
+import org.jsoup.select.Elements;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+
+import static com.microsoft.hydralab.common.util.Const.NetworkMonitorConfig.*;
+
+public class AndroidNetworkMonitor implements NetworkMonitor {
+
+    protected DeviceDriver deviceDriver;
+    protected ADBOperateUtil adbOperateUtil;
+    protected DeviceInfo deviceInfo;
+    protected String rule;
+    protected File resultFolder;
+    protected Logger logger;
+
+    public AndroidNetworkMonitor(DeviceDriver deviceDriver, ADBOperateUtil adbOperateUtil, DeviceInfo deviceInfo, String rule, File resultFolder, Logger logger) {
+        this.deviceDriver = deviceDriver;
+        this.adbOperateUtil = adbOperateUtil;
+        this.deviceInfo = deviceInfo;
+        this.rule = rule;
+        this.resultFolder = resultFolder;
+        this.logger = logger;
+    }
+
+    @Override
+    public void start() {
+        // launch vpn
+        String command_launch = "adb shell am start";
+        command_launch += " -a com.microsoft.hydralab.android.client.vpn.START";
+        command_launch += " -n com.microsoft.hydralab.android.client/.MainActivity";
+        ShellUtils.execLocalCommandWithResult(command_launch, logger);
+        while (true) {
+            ThreadUtils.safeSleep(1000);
+            boolean clicked = grantPermissionClick();
+            if (!clicked) {
+                break;
+            }
+        }
+
+        // start vpn
+        String command_start = "adb shell am start";
+        command_start += " -a com.microsoft.hydralab.android.client.vpn.START";
+        command_start += " -n com.microsoft.hydralab.android.client/.MainActivity";
+        command_start += String.format(" --es \"apps\" \"%s\"", rule);
+        command_start += " --es \"output\" \"" + AndroidDumpPath + "\"";
+        ShellUtils.execLocalCommandWithResult(command_start, logger);
+        while (true) {
+            ThreadUtils.safeSleep(1000);
+            boolean clicked = grantPermissionClick();
+            if (!clicked) {
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void stop() {
+        // stop vpn
+        String command_stop = "adb shell am start";
+        command_stop += " -a com.microsoft.hydralab.android.client.vpn.STOP";
+        command_stop += " -n com.microsoft.hydralab.android.client/.MainActivity";
+        ShellUtils.execLocalCommandWithResult(command_stop, logger);
+        ThreadUtils.safeSleep(2000);
+
+        // pull result
+        String local_dump_path = resultFolder.getAbsolutePath() + DumpPath;
```

Comment:
You can reuse com.microsoft.hydralab.common.util.ADBOperateUtil#pullFileToDir to pull file, there is a retry mechanism in this method.

Reply:
fixed